### PR TITLE
document spork system with doxygen comments

### PR DESCRIPTION
**Pull request by nmarley**
This is an initial shot at an on-going WIP, which will add documentation via Doxygen comments to (especially) Dash-specific features in the codebase.

@UdjinM6 I will need some guidance on the rationale for `GetSignatureHash`, as you will see. 😉

### CODE REVIEW

File: src/spork.cpp [hunk 1]
```diff
@@ -316,6 +378,10 @@ bool CSporkMessage::Sign(const CKey& key, bool fSporkSixActive)
     return true;
 }
 
+/**
+ * CheckSignature will ensire the spork signature matches the provided public
```
Comment (nmarley) [inline src/spork.cpp]: s/ensire/ensure/

File: src/spork.cpp [hunk 2]
```diff
@@ -260,23 +308,37 @@ bool CSporkManager::SetPrivKey(const std::string& strPrivKey)
     }
 }
 
+/**
+ * ToString returns the string representations of the SporkManager.
```
Comment (nmarley) [inline src/spork.cpp]: "representations" should be singular

**Comment by nmarley** [comment]
> May be it is better to place method descriptions in header file and to place only some implementation-specific notes in cpp file?

Thanks for the suggestion. I added them with the implementations because I thought that made more sense, but maybe the actual C++ way is to add them in the header files. Since the classes were declared in header files, that's where I added those class comments. Technically the methods are "declared" in the class definition (e.g. the function prototypes).

I guess the big questions (for me) are:

1. What is C++ convention for this?
2. Where are the comments more helpful?

~~It's possible doxygen also cares where these go.~~ Nope, looks like doxygen doesn't care. I see various opinions online. Wondering what others think.

**Comment by PastaPastaPasta** [comment]
> What is C++ convention for this?

Doesn't seem like there is one that I can find. 

Found this that makes sense (just some guys opinion):
* Document how to use the function in the header file, or more accurately close to the declaration.
* Document how the function works (if it's not obvious from the code) in the source file, or more accurately, close to the definition.

**Comment by nmarley** [comment]
@gladcow The more I look over DashCore code, I'm realizing that I don't actually look in the header files. And in my IDE when I click to go to the function, it's taking me to the implementation (e.g. the cpp file), not the header declaration (I use CLion).

Personally I don't think it makes since to stick the documentation in the header files where I'm not even looking at the file. The comments make the most sense to me where the method is defined (not declared).

**Comment by nmarley** [comment]
@PastaPastaPasta I'm almost certain your random guy's opinion came from StackExchange, which I also found yesterday as well, while searching conventions. I also found others which differ, also on Stack Exchange, such as this one (which I tend to agree with):

> Put the documentation where people will read and write it as they are using and working on the code.
> 
> Class comments go in front of classes, method comments in front of methods.
> 
> That is the best way to make sure things are maintained. It also keeps your header files relatively lean and avoids the touching issue of people updating method docs causing headers to be dirty and triggering rebuilds. I have actually known people use that as an excuse for writing documentation later!

File: src/spork.h [hunk 3]
```diff
@@ -75,15 +88,41 @@ class CSporkMessage
         }
     }
 
+    /**
+     * GetHash returns the double-sha256 hash of the serialized spork message. The
```
Comment (UdjinM6) [inline src/spork.h]: `The signature field is not included in the serialization.` this will not be true after #2288

File: src/spork.h [hunk 4]
```diff
@@ -75,15 +88,41 @@ class CSporkMessage
...
+     * signature field is not included in the serialization.
+     */
     uint256 GetHash() const;
+
+    /**
+     * GetSignatureHash is currently an alias for GetHash, and I'm not sure why it
```
Comment (UdjinM6) [inline src/spork.h]: >I'm not sure why

`The signature field is not included in the serialization.` - that's why in this case ;)

File: src/spork.h [hunk 5]
```diff
@@ -120,23 +159,84 @@ class CSporkManager
...
+
+    /**
+     * UpdateSpork is used by the spork RPC command to set a new spork value, sign
+     * and broadcast the spork message.
+     */
     bool UpdateSpork(int nSporkID, int64_t nValue, CConnman& connman);
 
+    /**
+     * IsSporkActive returns a bool for "on/off" sporks, which should be used to
+     * determine whether the spork can be considered active or not.
```
Comment (UdjinM6) [inline src/spork.h]: Should probably clarify that this is only true for time-based sporks (which are majority), for pure value-based sporks like SPORK_5_INSTANTSEND_MAX_VALUE this function makes no sense and should not be used.

**Comment by nmarley** [comment]
Fixed `Get*Hash` methods and added comment for `CheckAndRemove`, so that should be all (for now). I will start on the post-2288 work and rebase once that gets merged in.

**Comment by nmarley** [comment]
Added comments for new methods introduced in #2288, squashed & rebased to resolve conflicts. Please re-review.

**Pull request merged**